################################################################################
SHORT DESCRIPTION: 
################################################################################
Miscellaneous patches and fixes from the original OpenXT blktap.git repo.

################################################################################
LONG DESCRIPTION: 
################################################################################
The original blktap.git repo contained a mirror of blktap2 with years worth of
work directly checked in. Among this work were numerous fixes that are valid.

################################################################################
CHANGELOG 
################################################################################
Patch formed from commits by: Ross Philipson, philipsonr@ainfosec.com, 3/22/2016
Various original authors of the commits.
027f5abb706078b1f9fd49279f0d207849cd5e4b
5c6f405d11c3bf5fabf705c1ec35a1031e02a555
df2e16c7c6af2cebdef72d541ae192cbf6a5b55f

################################################################################
REMOVAL 
################################################################################
Can be removed at such a time as all the fixes are upstreamed.

################################################################################
UPSTREAM PLAN
################################################################################
Most of these patches can be upstreamed.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################

################################################################################
PATCHES 
################################################################################
Index: xen-4.3.4/tools/blktap2/vhd/lib/vhd-util-read.c
===================================================================
--- xen-4.3.4.orig/tools/blktap2/vhd/lib/vhd-util-read.c
+++ xen-4.3.4/tools/blktap2/vhd/lib/vhd-util-read.c
@@ -461,8 +461,10 @@ vhd_test_bitmap(vhd_context_t *vhd, uint
 
 		if (vhd->bat.bat[blk] == DD_BLK_UNUSED)
 			bit = 0;
-		else
-			bit = vhd_bitmap_test(vhd, buf, blk);
+		else {
+			/* Switch to sector, seems more in line with what -i does */
+			bit = vhd_bitmap_test(vhd, buf, sec);
+		}
 
 	print:
 		printf("block %s: ", conv(hex, blk));
@@ -510,8 +512,8 @@ vhd_test_batmap(vhd_context_t *vhd, uint
 
 	for (i = 0; i < count; i++) {
 		cur = block + i;
-		fprintf(stderr, "batmap for block %s: %d\n", conv(hex, cur),
-			vhd_batmap_test(vhd, &vhd->batmap, cur));
+		printf("batmap for block %s: %d\n", conv(hex, cur),
+		       vhd_batmap_test(vhd, &vhd->batmap, cur));
 	}
 
 	return 0;
Index: xen-4.3.4/tools/blktap2/vhd/lib/libvhd.c
===================================================================
--- xen-4.3.4.orig/tools/blktap2/vhd/lib/libvhd.c
+++ xen-4.3.4/tools/blktap2/vhd/lib/libvhd.c
@@ -2194,7 +2194,7 @@ vhd_write_block(vhd_context_t *ctx, uint
 	if (block >= ctx->bat.entries)
 		return -ERANGE;
 
-	if ((unsigned long)data & ~(VHD_SECTOR_SIZE -1))
+	if ((unsigned long)data & (VHD_SECTOR_SIZE -1))
 		return -EINVAL;
 
 	blk  = ctx->bat.bat[block];
Index: xen-4.3.4/tools/blktap2/drivers/tapdisk-image.c
===================================================================
--- xen-4.3.4.orig/tools/blktap2/drivers/tapdisk-image.c
+++ xen-4.3.4/tools/blktap2/drivers/tapdisk-image.c
@@ -86,10 +86,12 @@ tapdisk_image_free(td_image_t *image)
 int
 tapdisk_image_check_td_request(td_image_t *image, td_request_t treq)
 {
-	int rdonly;
+	int rdonly, err;
 	td_driver_t *driver;
 	td_disk_info_t *info;
 
+	err = -EINVAL;
+
 	driver = image->driver;
 	if (!driver)
 		return -ENODEV;
@@ -100,8 +102,10 @@ tapdisk_image_check_td_request(td_image_
 	if (treq.op != TD_OP_READ && treq.op != TD_OP_WRITE)
 		goto fail;
 
-	if (treq.op == TD_OP_WRITE && rdonly)
+	if (treq.op == TD_OP_WRITE && rdonly) {
+		err = -EPERM;
 		goto fail;
+	}
 
 	if (treq.secs <= 0 || treq.sec + treq.secs > info->size)
 		goto fail;
@@ -109,10 +113,10 @@ tapdisk_image_check_td_request(td_image_
 	return 0;
 
 fail:
-	ERR(-EINVAL, "bad td request on %s (%s, %"PRIu64"): %d at %"PRIu64,
+	ERR(err, "bad td request on %s (%s, %"PRIu64"): %d at %"PRIu64,
 	    image->name, (rdonly ? "ro" : "rw"), info->size, treq.op,
 	    treq.sec + treq.secs);
-	return -EINVAL;
+	return err;
 
 }
 
@@ -121,13 +125,14 @@ tapdisk_image_check_ring_request(td_imag
 {
 	td_driver_t *driver;
 	td_disk_info_t *info;
-	int i, psize, rdonly;
+	int i, err, psize, rdonly;
 	uint64_t nsects, total;
 
 	driver = image->driver;
 	if (!driver)
 		return -ENODEV;
 
+	err = -EINVAL;
 	nsects = 0;
 	total  = 0;
 	info   = &driver->info;
@@ -138,8 +143,10 @@ tapdisk_image_check_ring_request(td_imag
 	    req->operation != BLKIF_OP_WRITE)
 		goto fail;
 
-	if (req->operation == BLKIF_OP_WRITE && rdonly)
+	if (req->operation == BLKIF_OP_WRITE && rdonly) {
+		err = -EPERM;
 		goto fail;
+	}
 
 	if (!req->nr_segments || req->nr_segments > MAX_SEGMENTS_PER_REQ)
 		goto fail;
@@ -162,8 +169,8 @@ tapdisk_image_check_ring_request(td_imag
 	return 0;
 
 fail:
-	ERR(-EINVAL, "bad request on %s (%s, %"PRIu64"): id: %"PRIu64": %d at %"PRIu64,
+	ERR(err, "bad request on %s (%s, %"PRIu64"): id: %"PRIu64": %d at %"PRIu64,
 	    image->name, (rdonly ? "ro" : "rw"), info->size, req->id,
 	    req->operation, req->sector_number + total);
-	return -EINVAL;
+	return err;
 }
Index: xen-4.3.4/tools/blktap2/drivers/tapdisk-vbd.c
===================================================================
--- xen-4.3.4.orig/tools/blktap2/drivers/tapdisk-vbd.c
+++ xen-4.3.4/tools/blktap2/drivers/tapdisk-vbd.c
@@ -1181,12 +1181,25 @@ tapdisk_vbd_check_queue(td_vbd_t *vbd)
 	return 0;
 }
 
+static int
+tapdisk_vbd_request_should_retry(td_vbd_request_t *vreq)
+{
+	switch (vreq->error) {
+	case -EPERM:
+	case -ENOSYS:
+		return 0;
+	default:
+		return 1;
+	}
+}
+
 void
 tapdisk_vbd_complete_vbd_request(td_vbd_t *vbd, td_vbd_request_t *vreq)
 {
 	if (!vreq->submitting && !vreq->secs_pending) {
 		if (vreq->status == BLKIF_RSP_ERROR &&
 		    vreq->num_retries < TD_VBD_MAX_RETRIES &&
+		    tapdisk_vbd_request_should_retry(vreq) &&
 		    !td_flag_test(vbd->state, TD_VBD_DEAD) &&
 		    !td_flag_test(vbd->state, TD_VBD_SHUTDOWN_REQUESTED))
 			tapdisk_vbd_move_request(vreq, &vbd->failed_requests);
@@ -1394,13 +1407,17 @@ tapdisk_vbd_issue_request(td_vbd_t *vbd,
 
 #if 0
 	err = tapdisk_vbd_check_queue(vbd);
-	if (err)
+	if (err) {
+		vreq->error = err;
 		goto fail;
+	}
 #endif
 
 	err = tapdisk_image_check_ring_request(image, req);
-	if (err)
+	if (err) {
+		vreq->error = err;
 		goto fail;
+	}
 
 	for (i = 0; i < req->nr_segments; i++) {
 		nsects = req->seg[i].last_sect - req->seg[i].first_sect + 1;
